src/modules/control.cpp

```diff
@@ -2,60 +2,117 @@
+
 // Public Variables
 struct control control = {
   .state = CONTROL_IDLE
 };
 
 // Private Variables
 static struct pt controlThread;
+static struct timer breathTimer;
 static struct timer controlTimer;
 
-static unsigned int motorCompressionDistance;
-static unsigned int motorCompressionDuration;
+static uint16_t totalMotorCompressionDistance;
+static uint16_t totalMotorCompressionDuration;
+
+static uint16_t targetVolume;
+static uint32_t totalBreathTime;
+static uint32_t targetInhalationTime;
+
+static uint32_t measuredInhalationTime;
+static uint32_t measuredExhalationTime;
+
+// Mid-level volume control function
+static void updateVolumeControl(uint16_t* distance, uint16_t* duration)
+{
+  // TODO: fill-in this function
+}
 
 static PT_THREAD(controlThreadMain(struct pt* pt))
 {
   PT_BEGIN(pt);
   
   while (1) {
     if (control.state == CONTROL_IDLE) {
+      // Reset any parameters
+      control.breathCount = 0;
+      control.ieRatioMeasured = 0;
+      control.respirationRateMeasured = 0;
+      
       // Wait for the parameters to enter the run state before
-      if (parameters.startVentilation) {
-        control.state = CONTROL_BEGIN_INHALATION;
-      }
+      PT_WAIT_UNTIL(pt, parameters.startVentilation);
+      
+      control.state = CONTROL_BEGIN_INHALATION;
       
     } else if (control.state == CONTROL_BEGIN_INHALATION) {
-      if (parameters.ventilationMode == VENTILATOR_MODE_VC) {
-        // TODO: Calculate breathe parameters and motor control
-        motorCompressionDistance = 0;
-        motorCompressionDuration = 0;
-        // TODO: Error check
-        motorHalBegin(MOTOR_HAL_DIRECTION_INHALATION, motorCompressionDistance, motorCompressionDuration);
-      } else {
-        // TODO: Implement Assist mode setup
-      }
+      // Collect all set points from parameters
+      totalBreathTime = (60 SEC) / parameters.respirationRateRequested;
+      targetInhalationTime = (parameters.ieRatioRequested * totalBreathTime) >> 8;
+      targetVolume = parameters.volumeRequested;
+      
+      // Initialize all 
+      totalMotorCompressionDistance = 0;
+      totalMotorCompressionDuration = 0;
+      measuredInhalationTime = 0;
+      measuredExhalationTime = 0;
+      
+      // Kickoff timers for monitoring breath/stages of breathing
+      timerHalBegin(&breathTimer, totalBreathTime);
+      timerHalBegin(&controlTimer, targetInhalationTime * INHALATION_OVERTIME);
+      
       control.state = CONTROL_INHALATION;
       
     } else if (control.state == CONTROL_INHALATION) {
-      if (parameters.ventilationMode == VENTILATOR_MODE_VC) {
-        PT_WAIT_UNTIL(pt, motorHalRun() != HAL_IN_PROGRESS);
+      uint16_t motorDistance = 0;
+      uint16_t motorDuration = 0;
+  
+      // If in a volume controlled mode, run the Mid-level volume controller
+      if ((parameters.ventilationMode == VENTILATOR_MODE_VC) ||
+          (parameters.ventilationMode == VENTILATOR_MODE_AC)) {
```

Comment:
nit: `ventilationMode` vs. `VENTILATOR_MODE_*`

Consider switching definitions to `VENTILATION_MODE_*`.

Reply:
Yeah, modes need to be cleaned up anyway -- still waiting to hear back from Saurabh about it